### DISCUSSION TITLE
Single Field Winston Lutz Idea

### DISCUSSION THREAD

**Discussion by ethanio12345** [jrkerns/pylinac#441] [Ideas]
Hi there,

Had an idea for the single field WL test and just wanted to soundboard it as I'm not really sure how to code it within the Pylinac framework. Y'all may have already thought of it and have reasons against it. I'm fairly new to working on big projects (and this is my first suggestion ever...am a medphys reggie), so am welcome to feedback. 

If we had a large, symmetric, CAX centred, MLC defined field, we should be able to simulate smaller sub-fields and not lose significant accuracy in the positioning? Assuming the MLCs are visible in an image, and we know the nominal positions, then we should be able to calculate, fairly accurately, the positions of other points on an EPID? I.e., if we know/trust EPID readouts from the DICOM (offsets/roll/etc), then I'm wondering if could we apply a mask around where we expect each BB to be in the image, which breaks the problem down into a ``multi field'' problem, which you've already solved with the multi target implementation? 

Here's my algorithm thoughts 

- take large MLC defined field to give us field centre. 
- apply a binary mask (or just crop) near the ideal BB position (arbitrary field size, could also use Gaussian filter if we want to simulate field edge)
- Apply WL searching algorithm to search for single BB within sub field
- Repeat for all combinations of BB positions and gantry/col/couch
- Calculate idealised BB projection and compare measured position to expected position
- Summarise and work tech magic.

I've got some very basic code working for an image of MPC where I read the XIM, crop, thresh, openCV to find contour as a proof of concept. You can use a simple one-liner for the BB projections, as they are just an Euler rotation using extrinsic coords, which you can pull from ``scipy.spatial.transform import Rotation''. This allows calculating gantry, coll and couch pretty easily...you just need to make sure the rotations match (e.g. rotation of phantom is inverse of gantry)

Just an idea/test concept. Love this project and thanks for all the work!

**Comment by jrkerns** [comment]
Hey, thanks for bringing this up. 

To be honest, I'm a bit confused about what you're actually proposing. Are you wanting to do something like MPC where the displacement from the ideal positions is reversed to get the gantry/coll/couch positional error? If so I don't think you'll be able to get away with a single image for that!

If you are referring more to an MPC-like analysis where there are multiple BBs but one large field vs the multiple subfields of the SNC multi-target WL, it's in our queue 😉. 

I was not aware of the `scipy.spatial.transform` module. I might be able to simplify some of my code with that. I'm not very good with transforms so I'll have to do some research.

If you restate your problem and what output you want I might be able to get more specific.

Thanks

**Reply by ethanio12345** [reply]
Thanks for the reply!

Thinking about the latter one (sorry if it wasn't clear). Definitely agree there wouldn't be enough info in a single image to get away with the first! 😂

I did read it was in the queue, but wasn't sure how far along it was. If y'all already have a good idea for how to do it, then ignore me (as I just wanted to pose an idea for it) and continue the good work!

Scipy has loads of great stuff (I'm a fan of not reinventing the wheel). https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_euler.html is a good start. Look into the lowercase "xyz" for external What I meant r.e. rotation was more confusing than it needed to be. For my first attempt I did "a dumb" and rotated my phantom instead of the gantry and wondered why my gantry angles didn't match what I thought...doh. Reading through the code again, the bb gantry projection function should just be fairly simple. You already import part of scipy, so it shouldn't be a deal breaker. 😂

Anyway, thanks for the ear! Appreciate your time.

**Reply by jrkerns** [reply]
I haven't started the MPC-like analysis, but the hard part was largely completed in the new multi-field multi-target module. We don't have a deadline but there seems to be a big general interest in independently recomputing MPC results. I also had to build the XIM reader since MPC doesn't save DICOM files. From my perspective MPC is extremely accurate so I'm not sure if there's a big advantage. But hey, the people ask...

However, I'm definitely interested in this rotation transformation. I already use scipy all over in pylinac, just not that module, so no worries about the dependency. 

I'll definitely take a look and thanks for the tip. I'm happy to accept PRs or utilize code from a prototype. For a refactor like the gantry projection, these are great because it's already backed by tests so I'll know if I screwed it up! 🙃

**Reply by jrkerns** [reply]
ChatGPT came in clutch here with explanations and numerous examples. 
![image](https://user-images.githubusercontent.com/4803336/233687094-d19cd521-77f3-4aa6-bcec-4a6486ba25ff.png)

**Reply by ethanio12345** [reply]
I have a lot of trust in MPC (which I think literature supports), but the ability to go back and recalculate certain things (even like a proper flatness/symmetry rather than their single high/low point check) gives it a lot more weight from an independent sense. 😊

Oh man, that one is awesome!  ChatGPT looks super helpful. The main reason I chose the Euler angles is more that it conceptually matches a Linac in my thoughts as a step by step (set something up perfectly to iso, rotate one axos, rotate another, rotate another)...still, if you find one you like more, go ahead. Am happy I could suggest/help in some small manner. I could flick some very messy prototype code your way if you wanted...where is the best place for it?

**Reply by jrkerns** [reply]
Fork pylinac, make a new branch and link it here.

**Reply by ethanio12345** [reply]
Ok. So, back to work and on lunch...but silly proxies and hospital networks. Managed to get there. Tested with your Offset BBs code in the docs for a 1mm offset and it's giving the same output.

Basically, the changes are at the bottom of the WL file, just added a new function, and changing the shifts in the _nominal_point. So, represent the BB as a vector around iso, rotate the BB position vector using scipy from_euler, place bb at iso (i.e. add iso to up/down along BEV), use new up/down in BEV to find magnification to MV panel return the projection onto the imager. I did have it in my prototype code to be vectorised, so you could just pass a whole series of BBs, but redid it to keep it for a singular bb for the pylinac.

Hoping this is what you meant - https://github.com/ethanio12345/pylinac/tree/ethanio12345-patch-1

**Reply by jrkerns** [reply]
That works for me. I'm off for the next few days and I didn't dig deep yet but if the tests pass this will be great! I'll get back to you in a bit. Thanks!

**Reply by ethanio12345** [reply]
No worries mate! Enjoy the time off! When you get back, take this commit if you can (568e8058e2ad7a7c62e6e783e637154179426684)...I added a test file in the latest one, so probably not necessary...not very familiar with unit tests...should look into them. :)

**Reply by jrkerns** [reply]
Okay, so from what I can tell so far, this works 🎉 It should be more concise than my current ad hoc implementations: https://github.com/jrkerns/pylinac/blob/master/pylinac/winston_lutz.py#L1943-L1973. However, I have quite a few tests written up for the existing functions already https://github.com/jrkerns/pylinac/blob/master/tests_basic/test_winstonlutz.py#L35-L271 (wasted effort in retrospect 😢). I'd like to replace the existing ad hoc functions with a function similar to this one. I'll need to write some tests for it but this should work out nicely. I want to also use a similar function to calculate the roll, pitch, and yaw for multi-bb analysis so that a 4 or 6D shift can be applied to the couch to place the phantom perfectly. 

Separately, we update pylinac in Bitbucket (what Radformation uses for VCS) first and then it's mirrored here on Github. All that to say that when I replace the functions with what you have here I'll have to copy/paste it but will link your contribution and this issue in the docstring and release notes, but I can't merge PRs ATM.

**Reply by ethanio12345** [reply]
That case with the couch and multi-bb sounds like a great use case of that ``align vectors'' you found in scipy! 

Not being able to merge PRs sounds interesting, but no worries there! Happy to leave it in your capable hands. Guessing it's time to close this discussion. Until next time! :)

**Reply by jrkerns** [reply]
Lemme see if there's some git magic I can do...

**Reply by ethanio12345** [reply]
Only for the purpose of making a dad joke (and not to be pushy), you should ``git'' onto it. :P

**Reply by jrkerns** [reply]
push-y 🙃

**Reply by ethanio12345** [reply]
https://media.tenor.com/UladBrWhuJ8AAAAM/bowing-thank-you.gif

**Reply by jrkerns** [reply]
Okay yeah I'm stupid, I can get it in no problem. I'm not quite ready to have it replace the existing functions though before I write some tests. If you revert the changes inside `_nominal_point` (i.e. just the function exists w/ no usages) and make a PR, then I'll merge it. That way the function is there and ready when I get to it and you'll also get git blame 😄

**Reply by ethanio12345** [reply]
Ok. Think I've done it. Made a dumb, tried to correct the dumb, realised it wasn't really a dumb and I was just confusing myself. PR made.
